### PR TITLE
basic working

### PR DESCRIPTION
**Pull request by Lokathor**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,25 @@
...
+<!-- [![AppVeyor](https://ci.appveyor.com/api/projects/status/r7am2u1vw9m7qrge?svg=true)](https://ci.appveyor.com/project/Lokathor/hektor) -->
+[![crates.io](https://img.shields.io/crates/v/proc-spirv.svg)](https://crates.io/crates/proc-spirv)
+[![docs.rs](https://docs.rs/proc-spirv/badge.svg)](https://docs.rs/proc-spirv/)
+
 # proc-spirv
-procedural macro your SPIR-V creation
+
+A proc-macro crate to get your shaders compiled into SPIR-V at compile time.
+
+Currently we use the [shaderc](https://docs.rs/shaderc) crate but that's not
```
Comment (memoryruins) [inline README.md]:
```suggestion
Currently, we use the [shaderc](https://docs.rs/shaderc) crate, but that's not
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,25 @@
...
+considered a "stable" part of the API. Anything that turns GLSL into SPIR-V
+would work in the long term. If someone comes up with a solid pure-rust GLSL to
+SPIR-V compiler we could easily switch to that.
```
Comment (memoryruins) [inline README.md]:
```suggestion
SPIR-V compiler, we could easily switch to that.
```

File: README.md [hunk 3]
```diff
@@ -1,2 +1,25 @@
...
+
+Probably no one will do that because people are lazy.
+
+## When To Use This
+
+If you want to have some SPIR-V shader code in your program but you _do not_
+need to have your program be compiling new shaders at runtime, you can use this
+crate's proc-macro instead to do all the processing at compile time only. This
+makes your final binary be smaller, because it's not carrying around a SPIR-V
```
Comment (memoryruins) [inline README.md]:
```suggestion
makes your final binary smaller, because it's not carrying around a SPIR-V
```

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  let filename_string = filename.value();
+  let file_handle = match File::open(&filename_string) {
+    Ok(f) => f,
+    Err(e) => panic!("Failed to open the file specified: {:?}", e),
+  };
+  let mut buf_reader = BufReader::new(file_handle);
+  let mut file_contents = String::new();
+  match buf_reader.read_to_string(&mut file_contents) {
+    Ok(_) => (),
+    Err(e) => panic!("Failed to read the file contents: {:?}", e),
```
Comment (memoryruins) [inline src/lib.rs]:
```rust
buf_reader.read_to_string(&mut file_contents)
  .unwrap_or_else(|e| panic!("Failed to read the file contents: {:?}", e));
```
or
```rust
if let Err(e) = buf_reader.read_to_string(&mut file_contents) {
    panic!("Failed to read the file contents: {:?}", e)
}
```

File: src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  let SPIRVFromFile {
+    kind,
+    filename,
+    entry_point,
+  } = parse_macro_input!(input as SPIRVFromFile);
+
+  let filename_string = filename.value();
+  let file_handle = match File::open(&filename_string) {
+    Ok(f) => f,
+    Err(e) => panic!("Failed to open the file specified: {:?}", e),
```
Comment (memoryruins) [inline src/lib.rs]:
```rust
let file_handle = File::open(&filename_string)
  .unwrap_or_else(|e| panic!("Failed to open the file specified: {:?}", e));

File: src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  let compile_result = compiler.compile_into_spirv(
+    source_text,
+    shader_kind,
+    input_file_name,
+    entry_point_name,
+    additional_options,
+  );
+
+  let artifact = match compile_result {
+    Ok(artifact) => artifact,
```
Comment (memoryruins) [inline src/lib.rs]:
```rust
let artifact = compile_result
  .unwrap_or_else(|e| panic!("SPIR-V compilation error: {}", e));
```

File: src/lib.rs (hunk 4 above)
Comment (Lokathor) [reply src/lib.rs]: we'll take option 2 for this one

Comment (Lokathor) [reply src/lib.rs]: WAIT OPTION 1 MATCHES THE NEXT FIX BETTER

File: src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    input_file_name,
+    entry_point_name,
+    additional_options,
+  );
+
+  let artifact = compile_result.unwrap_or_else(|e| panic!("SPIR-V compilation error: {}", e));
+
+  let out_string = format!("&{:?}", artifact.as_binary_u8());
+
+  TokenStream::from_str(&out_string)
```
Comment (memoryruins) [inline src/lib.rs]: this one

Comment (Lokathor) [reply src/lib.rs]: got got

**Pull request merged**